pkg/storage/stores/shipper/gateway_client.go

```diff
@@ -136,6 +170,56 @@ func (s *GatewayClient) doQueries(ctx context.Context, queries []chunk.IndexQuer
 	return nil
 }
 
+// ringModeDoQueries prepares an index query to be sent to the Index Gateway, and then sends it
+// using the clientDoQueries implementation.
+//
+// The preparation and sending phase includes:
+// 1. Extracting the tenant name from the query.
+// 2. Fetching different Index Gateway instances assigned to the extracted tenant.
+// 3. Iterating in parallel over all fetched Index Gateway instances, getting their gRPC connections
+//  from the pool and invoking clientDoQueries using their client.
+func (s *GatewayClient) ringModeDoQueries(ctx context.Context, gatewayQueries []*indexgatewaypb.IndexQuery, queryKeyQueryMap map[string]chunk.IndexQuery, callback chunk.QueryPagesCallback) error {
+	// TODO: should we keep iterating if an error occur?
+	for _, q := range gatewayQueries {
+		bufDescs, bufHosts, bufZones := ring.MakeBuffersForGet()
+
+		splittedHashed := strings.Split(q.HashValue, ":")
+		if len(splittedHashed) < 2 {
+			level.Error(util_log.Logger).Log("msg", "query hash value in incorrect format", "hash_value", q.HashValue)
+			continue
+		}
+
+		// tenant is the second parameter.
+		tenantID := splittedHashed[1]
```

Comment:
You can get the `tenantID` from the context like this here https://github.com/grafana/loki/blob/6cc503484cb1c12d255ef110a7337f72224e174d/pkg/storage/store.go#L214
And the `tenantID` for a batch of queries would always be the same.

Reply:
great catch, thanks.

Reply:
Done. PTAL?